File: payment-service/strategies/NetBankingProcessor.js
import { PaymentProcessor } from './PaymentProcessor.js';
import { v4 as uuidv4 } from 'uuid';

export class NetBankingProcessor extends PaymentProcessor {
    constructor() {
        super();
        this.PAYMENT_METHOD = 'NETBANKING';
        this.PROCESSING_DELAY_MS = 6000; // NetBanking is slowest
    }

    async process(transaction) {
        console.log(`Processing NetBanking payment for transaction: ${transaction.id}`);
        
        return new Promise((resolve) => {
            setTimeout(() => {
                // Simulate success/failure (80% success rate for NetBanking)
                const isSuccess = Math.random() < 0.8;
                
                if (isSuccess) {
                    const netBankingTransactionId = 'NB' + uuidv4().substring(0, 10).toUpperCase();
                    console.log(`NetBanking payment successful for transaction: ${transaction.id} with NB ID: ${netBankingTransactionId}`);
                    resolve(PaymentProcessor.success(netBankingTransactionId, 'NetBanking payment completed successfully'));
                } else {
                    const failureReason = this.getRandomNetBankingFailureReason();
                    console.warn(`NetBanking payment failed for transaction: ${transaction.id} - ${failureReason}`);
                    resolve(PaymentProcessor.failure(failureReason));
                }
            }, this.PROCESSING_DELAY_MS);
        });
    }

    getPaymentMethod() {
        return this.PAYMENT_METHOD;
    }

    canProcess(transaction) {
        return super.canProcess(transaction) && transaction.amount >= 1; // Minimum NetBanking amount
    }

    getRandomNetBankingFailureReason() {
        const reasons = [
            'Bank server maintenance',
            'Session timeout',
            'Invalid credentials',
            'Account temporarily blocked',
            'Daily transaction limit reached',
            'Technical error at bank'
        ];
        return reasons[Math.floor(Math.random() * reasons.length)];
    }
}

File: backend/src/main/java/com/upi/gateway/backend/strategy/impl/NetBankingProcessor.java
package com.upi.gateway.backend.strategy.impl;

import com.upi.gateway.backend.model.Transaction;
import com.upi.gateway.backend.strategy.PaymentProcessor;
import com.upi.gateway.backend.strategy.PaymentResult;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ThreadLocalRandom;

@Component
@Slf4j
public class NetBankingProcessor implements PaymentProcessor {
    
    private static final String PAYMENT_METHOD = "NETBANKING";
    private static final int PROCESSING_DELAY_MS = 6000; // NetBanking is slowest
    
    @Override
    public CompletableFuture<PaymentResult> process(Transaction transaction) {
        log.info("Processing NetBanking payment for transaction: {}", transaction.getId());
        
        return CompletableFuture.supplyAsync(() -> {
            try {
                // Simulate NetBanking processing time
                Thread.sleep(PROCESSING_DELAY_MS);
                
                // Simulate success/failure (80% success rate for NetBanking)
                boolean isSuccess = ThreadLocalRandom.current().nextDouble() < 0.8;
                
                if (isSuccess) {
                    String netBankingTransactionId = "NB" + UUID.randomUUID().toString().substring(0, 10).toUpperCase();
                    log.info("NetBanking payment successful for transaction: {} with NB ID: {}", 
                            transaction.getId(), netBankingTransactionId);
                    return PaymentResult.success(netBankingTransactionId, "NetBanking payment completed successfully");
                } else {
                    String failureReason = getRandomNetBankingFailureReason();
                    log.warn("NetBanking payment failed for transaction: {} - {}", transaction.getId(), failureReason);
                    return PaymentResult.failure(failureReason);
                }
                
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                log.error("NetBanking processing interrupted for transaction: {}", transaction.getId());
                return PaymentResult.failure("Processing interrupted");
            }
        });
    }
    
    @Override
    public String getPaymentMethod() {
        return PAYMENT_METHOD;
    }
    
    @Override
    public boolean canProcess(Transaction transaction) {
        return PAYMENT_METHOD.equalsIgnoreCase(transaction.getPaymentMethod()) &&
               transaction.getAmount().doubleValue() >= 1; // Minimum NetBanking amount
    }
    
    private String getRandomNetBankingFailureReason() {
        String[] reasons = {
            "Bank server maintenance",
            "Session timeout",
            "Invalid credentials",
            "Account temporarily blocked",
            "Daily transaction limit reached",
            "Technical error at bank"
        };
        return reasons[ThreadLocalRandom.current().nextInt(reasons.length)];
    }
}